### DISCUSSION TITLE
Crash when shutdown the app

### DISCUSSION THREAD

**Discussion by pengweichu** [uNetworking/uWebSockets#1121] [Q&A] [answered]
Hi, sorry for too many questions, but I‘m struggling to shut down the app, first I got a crash when shutdown my app, then I copy the sample code at the bottom of this issue #1050 from @vidcentum (thank you :) ), but have the same issue, attached the screenshot, I'm using windows 10 (version 2004, 64bit), VS2019, 16.7.7.

In that sample code, I've just added a function called closeListenSocket, then when I want to shut down the app, I called this function on each thread listen socket, the crash appears, I think it's my wrong usage? Thanks

```
    void closeListenSocket()
    {
        us_listen_socket_close(0, listen_socket_);
    }
```

```
/* Main */
int main() {

    workers.resize(std::thread::hardware_concurrency());

    std::transform(workers.begin(), workers.end(), workers.begin(), [](worker_t& w) {
        w.thread_ = std::make_shared<std::thread>([&w]() {
            /* create uWebSocket worker and capture uWS::Loop, uWS::App objects. */
            w.work();
            });
        return w;
        });

    std::this_thread::sleep_for(std::chrono::seconds(5));

    std::for_each(workers.begin(), workers.end(), [](worker_t& w) {
        w.closeListenSocket();   // -> Added this line to close the listen socket in each thread
        w.thread_->join();
        });

    std::cout << "done" << std::endl;
    return 0;
}
```

<img width="1275" alt="crash" src="https://user-images.githubusercontent.com/12697283/98568338-f059ad80-22eb-11eb-9fa6-5a47b2609fd3.png">

**Comment by pengweichu** [comment]
in us_internal_loop_unlink function, we maybe need to check the context->prev is empty or not before unlink:

```
void us_internal_loop_unlink(struct us_loop_t *loop, struct us_socket_context_t *context) {
    if (loop->data.head == context) {
        loop->data.head = context->next;
        if (loop->data.head) {
            loop->data.head->prev = 0;
        }
    } else {
        context->prev->next = context->next; // Here, need to check the prev empty or not
        if (context->next) {
            context->next->prev = context->prev;
        }
    }
}
```

**Comment by ghost** [comment, marked as answer]
Well okay the manual doesn't mention that much about threading other than things are single threaded. But this is immensely important to follow. You are breaking this over and over. The only function that is thread safe is Loop::defer. All other functions are not thread safe.
